lib/slack/real_time/stores/loaded.rb

```diff
@@ -0,0 +1,80 @@
+          mpims: 'mpim'
+        }.freeze
+
+        attr_reader :caches
+
+        def initialize(_attrs, options = {})
+          @caches =
+            if options[:caches].is_a?(Array) || options[:caches].nil?
+              unknown_caches = options[:caches].to_a.map(&:to_sym) - CACHES
+              raise ArgumentError, "Unknown caches: #{unknown_caches.join(', ')}" unless unknown_caches.empty?
+
+              Set.new(options[:caches]).freeze
+            elsif options[:caches] == :all
+              Set.new(CACHES).freeze
+            else
+              raise ArgumentError, 'Unexpected value for option `caches`. Expects :all or an array of caches'
+            end
+          super
+        end
+
+        # The client has successfully connected to the server.
+        # @see https://api.slack.com/events/hello
+        # @see https://github.com/slack-ruby/slack-api-ref/blob/master/events/hello.json
+        on :hello do |_data, client|
+          if caches.include?(:teams)
+            team_info = client.web_client.team_info
+            team = Models::Team.new(team_info.team)
+            teams[team.id] = teams.key?(team.id) ? teams[team.id].merge(team) : team
+          end
+
+          if caches.include?(:users) || caches.include?(:bots)
+            client.web_client.users_list do |users_list|
+              users_list.members.each do |user_data|
+                if user_data.is_bot? && caches.include?(:bots)
+                  bot = Models::Bot.new(user_data)
+                  bots[bot.id] = bots.key?(bot.id) ? bots[bot.id].merge(bot) : bot
+                elsif user_data.is_user? && caches.include?(:users)
+                  user = Models::User.new(user_data)
+                  users[user.id] = users.key?(user.id) ? users[user.id].merge(user) : user
+                end
+              end
+            end
+          end
+
+          if caches.include?(:channels) || caches.include?(:groups) || caches.include?(:ims) || caches.include?(:mpims)
+            types = CONVERSATION_TYPES.slice(*caches).values.join(',')
+            client.web_client.conversations_list(types: types) do |conversations|
+              conversations.channels.each do |channel_data|
+                if channel_data.is_mpim?
+                  mpim = Models::Mpim.new(channel_data)
+                  mpims[mpim.id] = mpims.key?(mpim.id) ? mpims[mpim.id].merge(mpim) : mpim
+                elsif channel_data.is_im?
+                  im = Models::Im.new(channel_data)
+                  ims[im.id] = ims.key?(im.id) ? ims[im.id].merge(im) : im
+                elsif channel_data.is_group?
+                  group = Models::Group.new(channel_data)
+                  groups[group.id] = groups.key?(group.id) ? groups[group.id].merge(group) : group
+                elsif channel_data.is_channel?
+                  channel = Models::Channel.new(channel_data)
+                  channels[channel.id] = channels.key?(channel.id) ? channels[channel.id].merge(channel) : channel
```

Comment:
I'm conflicted about what to do with groups. I distinctly remember when I set up a bot several years ago that `channels` were effectively public channels and `groups` were private channels.

I finally found this explanation: https://api.slack.com/types/group
> A group object contains information about a private channel created before March 2021. The private channel's ID will begin with G, as private channels were once known as "private groups." [Conversation objects](https://api.slack.com/types/types/conversation) contain the information for private channels created after March 2021 (those with IDs that begin with C).

I'm not sure why they changed it, but my interpretation is that groups aren't created anymore.

So there are
public channels: `{ is_channel: true, is_group: false, is_private: false }`
private channels (created after March 2021): `{ is_channel: true, is_group: false, is_private: true }`
private channels (created before March 2021): `{ is_channel: false, is_group: true, is_private: true }`

It seems strange to me that groups are not labeled as `is_channel`, but they are effectively treated the same and `conversations_list(types: 'private_channel')` returns both private channel types in the same list.

I think having separate caches for the public channels and private channels is a lot more useful than a cache for all "channels" (both public and private) and a separate cache for just groups, but I'm not sure what to name them. Keeping them named `channels` and `groups` is convenient and probably makes sense to someone familiar with the pre-March 2021 behavior, but calling them `channels`/`public_channels` and `private_channels`, while more verbose, may be clearer for fresh users.

Reply:
Sigh, this stuff is wild :) I think we should remove any legacy behavior altogether, thus calling `conversations_list` with `public_channel` on startup by default, making it configurable, and storing a hash by `type` from the new API, so only buckets like `public_channel`, `private_channel`, `mpim`, `im` that match 100% what you'd get back from the API if you specified `type`. WDYT?

Reply:
I agree with matching the store's caches up with the types used by Slack (despite `public_channels` and `private_channels` being annoyingly long) 😬 

When you say
> storing a hash by type from the new API ... extensible for any future types Slack might add

are you thinking the stores only contain a single variable with a Hash that gets the types as keys? I prefer still having a separate Hash for each type like it has now. I think that's simpler and easier for dealing with the types as needed and the types are so fundamental that they rarely change (private groups becoming private channels is the only change in the last few years AFAIK).

And then I was also unsure about this
> thus calling `conversations_list` with `public_channel` on startup by default

since before you mentioned doing less by default. I guess if all of the extra initialization calls are configurable, there isn't a need for the separate `Loaded` store I set up and they could instead just be in the main `Store`. It's just a question of which caches should be fetched by default. Could be none, could be all, or could be ones we consider typically useful (like all except `mpims`). If `Starter` is the default store, then there's some flexibility in the default configuration for the standard `Store`.

Reply:
From the developer perspective I think I prefer to write `store.conversations.public_channel` vs. `store.public_channel`, because it matches the API, but that's not the hill I want to die on.

Agree on fetching nothing by default, I think you're right. As long as it's configurable.

Reply:
Ah I see what you were saying now with `store.conversations.public_channel`. My counter would be that the methods have changed more frequently than the types so I think it's important to decouple them. There have also been different methods that could fetch similar data and already we would be using a few different methods for getting team data vs user-like data vs channel-like data.

If that's an acceptable approach, I'll try to update this by today or tomorrow :)

Reply:
Acceptable!